Module1.FileSystemVisitor/FileSystemVisitor/FileSystemVisitor.cs

```diff
@@ -0,0 +1,179 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Project
+{
+    public class FileSystemVisitor
+    {
+        private Func<FileSystemInfo, bool> _func;
+
+        private bool _isSearchComplete = false;
+
+        private string _nextFileToSkip;
+
+        public event EventHandler<WorkEventArgs> WorkStarted;
+
+        public event EventHandler<WorkEventArgs> WorkFinished;
+
+        public event EventHandler<FileFoundEventArgs> FileFound;
+
+        public event EventHandler<FileFoundEventArgs> DirectoryFound;
+
+        public event EventHandler<FileFoundEventArgs> FilteredFileFound;
+
+        public event EventHandler<FileFoundEventArgs> FilteredDirectoryFound;
+
+        #region Constructors
+
+        public FileSystemVisitor()
+        {
+            _func = (x) => true;
+        }
+
+        public FileSystemVisitor(Func<FileSystemInfo, bool> func)
+        {
+            _func = func;
+        }
+
+        #endregion
+
+        public void CompleteSearch()
+        {
+            _isSearchComplete = true;
+        }
+
+        public void SkipFile(string fileName)
+        {
+            _nextFileToSkip = fileName;
+        }
+
+        public IEnumerable<string> GetFileList(string directoryPath)
+        {
+            OnWorkStarted(this, new WorkEventArgs("Work started."));
+
+            var directoryInfo = new DirectoryInfo(directoryPath);
+            foreach (var file in directoryInfo.GetFiles())
+            {
+                OnFileFound(this, new FileFoundEventArgs("File found.", file.FullName));
+
+                if (_func(file))
+                {
+                    OnFilteredFileFound(this, new FileFoundEventArgs("Filtered file found.", file.FullName));
+
+                    if (_isSearchComplete)
+                    {
+                        yield break;
+                    }
+
+                    if (file.FullName.Equals(_nextFileToSkip))
+                    {
+                        continue;
+                    }
+
+                    yield return file.FullName;
+                }
+            }
+
+            foreach (var directory in directoryInfo.GetDirectories())
+            {
+                OnDirectoryFound(this, new FileFoundEventArgs("Directory found.", directory.FullName));
```

Comment:
FileFoundEventAgs used in loop for directory, that's a bit confused

Reply:
done